### PR TITLE
arch: arm: replace main thread switching assembly to C using cmsis

### LINKED ISSUE #12878: ARM: Replace assembly code with CMSIS-based C code in ARM kernel_arch_func.h
**Is your enhancement proposal related to a problem? Please describe.**
We want to get rid of assembly code in ARM initialization code. We can very well use C code, and make use of CMSIS intrinsic functions wherever possible.

**Describe the solution you'd like**
Replace assembly code in ARM `kernel_arch_func.h` with C code.

**Describe alternatives you've considered**
Alternatively, we can keep the assembly code; it works ok, but it is hard to maintain and we have to manually ensure we respect and take into consideration the ARM calling convention.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by manochavikas**
use cmsis functions to avoid using direct assembly for main thread
switching.

This patch uses cmsis functions to:
	- set PSP(process stack pointer) to main stack
	- enable interrupts

Fixes #12878 

Signed-off-by: Vikas Manocha <vikas.manocha@st.com>

### CODE REVIEW

**Review by pizi-nordic — APPROVED** [review]

**Review by ioannisg — CHANGES_REQUESTED** [review]
Thanks for the patch. I think configure_mpu_stack_guard needs to be included (just browsed CI failure)

**Comment by ioannisg** [comment]
Also, this overlaps, in part with #12954, basically cancelling the required push/pop re-work in this file, since this will now be C code.

**Comment by manochavikas** [comment]
configure_mpu_stack_guard() is being called as before within
CONFIG_MPU_STACK_GUARD.
the only thing I removed was push/pop of {r2, LR}, not sure why it was
there, we never used these registers in code.

Cheers,
Vikas

**Comment by manochavikas** [comment]
Hi,

> On Feb 1, 2019, at 6:07 AM, Ioannis Glaropoulos <notifications@github.com> wrote:
> 
> configure_mpu_stack_guard() is being called as before within CONFIG_MPU_STACK_GUARD. the only thing I removed was push/pop of {r2, LR}, not sure why it was there, we never used these registers in code. Cheers, Vikas
> …
> On Fri, Feb 1, 2019 at 3:02 AM Ioannis Glaropoulos ***@***.***> wrote: @ioannisg requested changes on this pull request. Thanks for the patch. I think configure_mpu_stack_guard needs to be included (just browsed CI failure) — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or mute the thread.
> 
> Now it is C code, not assembly, I think it now requires including the declaration. But I am not sure ,pls, check. :)
> 

It makes sense, I will fix it in next pull request.

**Comment by manochavikas** [comment]
@ioannisg , please find the updated pull request fixing the missing the function prototype issue.

**Review by agross-oss — APPROVED** [review]

**Comment by ioannisg** [comment]
Could you please rebase, @manochavikas ?

**Comment by ioannisg** [comment]
Some function calls have been recently updated.

**Comment by manochavikas** [comment]
Done, Rebased to master.

**Review by ioannisg — APPROVED** [review]
Looks good, thank you very much @manochavikas .
I left a minor style comment.

Also: in the commit message body, could you add a line with:

Fixes #12878.

This would make #12878 automatically close, once the PR is merged.

**Comment by andrewboie** [comment]
> Also: in the commit message body, could you add a line with:
> 
> Fixes #12878.
> 
> This would make #12878 automatically close, once the PR is merged.

You can just edit the PR description to accomplish this, which I have done just now.

File: arch/arm/include/kernel_arch_func.h [hunk 1]
```diff
@@ -31,6 +31,7 @@ extern void _FaultInit(void);
 extern void _CpuIdleInit(void);
 #ifdef CONFIG_ARM_MPU
 extern void _arch_configure_static_mpu_regions(void);
+extern void _arch_configure_dynamic_mpu_regions(struct k_thread *thread);
 #endif
```
[the remark this answers is not in the document]
Comment (manochavikas) [reply arch/arm/include/kernel_arch_func.h]: :-), done in next pull, also added the PR

**Comment by manochavikas** [comment]
> > Also: in the commit message body, could you add a line with:
> > Fixes #12878.
> > This would make #12878 automatically close, once the PR is merged.
> 
> You can just edit the PR description to accomplish this, which I have done just now.

Thanks Andrew.

**Pull request merged**